File: mcp-code-intelligence-nodejs/src/graph/dependency-graph-service.ts
/**
 * KSA-155: Dependency Graph Service - BFS traversal on import relationships.
 * Supports outgoing (what does this file import?) and incoming (who imports this file?) queries.
 */

import * as path from 'path';
import Database from 'better-sqlite3';
import { FileResolver } from './file-resolver.js';

export interface DependencyNode {
  file: string;
  depth: number;
  importedSymbols: string[];
  isExternal: boolean;
}

export interface DependencyResult {
  root: string;
  direction: string;
  results: DependencyNode[];
  cycles: string[][];
  metadata: {
    totalNodes: number;
    maxDepthReached: number;
    truncated: boolean;
    queryTimeMs: number;
    externalCount: number;
  };
}

export class DependencyGraphService {
  private db: Database.Database;
  private fileResolver: FileResolver;

  constructor(db: Database.Database, fileResolver: FileResolver) {
    this.db = db;
    this.fileResolver = fileResolver;
  }

  /** Query dependency graph with direction and depth control. */
  query(
    file: string,
    direction: 'incoming' | 'outgoing' | 'both' = 'outgoing',
    depth: number = 1,
    includeExternal: boolean = false,
    limit: number = 50
  ): DependencyResult {
    const startTime = Date.now();
    const clampedDepth = Math.min(Math.max(depth, 1), 5);

    const resolved = this.fileResolver.resolveFile(file);
    if (!resolved) {
      return this.fileNotFoundResponse(file);
    }

    let results: DependencyNode[];
    let cycles: string[][];

    if (direction === 'both') {
      const outgoing = this.bfsTraversal(resolved, 'outgoing', clampedDepth, includeExternal, limit);
      const incoming = this.bfsTraversal(resolved, 'incoming', clampedDepth, includeExternal, limit);
      results = this.mergeResults(outgoing.results, incoming.results);
      cycles = [...outgoing.cycles, ...incoming.cycles];
    } else {
      const traversal = this.bfsTraversal(resolved, direction, clampedDepth, includeExternal, limit);
      results = traversal.results;
      cycles = traversal.cycles;
    }

    return {
      root: resolved,
      direction,
      results,
      cycles,
      metadata: {
        totalNodes: results.length,
        maxDepthReached: Math.min(clampedDepth, Math.max(...results.map(r => r.depth), 0)),
        truncated: results.length >= limit,
        queryTimeMs: Date.now() - startTime,
        externalCount: results.filter(r => r.isExternal).length,
      },
    };
  }

  private bfsTraversal(
    root: string,
    direction: 'incoming' | 'outgoing',
    maxDepth: number,
    includeExternal: boolean,
    limit: number
  ): { results: DependencyNode[]; cycles: string[][] } {
    const visited = new Set<string>([root]);
    const results: DependencyNode[] = [];
    const cycles: string[][] = [];
    const queue: Array<{ file: string; depth: number; path: string[] }> = [
      { file: root, depth: 0, path: [root] },
    ];

    while (queue.length > 0 && results.length < limit) {
      const { file: current, depth: currentDepth, path: currentPath } = queue.shift()!;
      if (currentDepth >= maxDepth) continue;

      const deps = direction === 'outgoing'
        ? this.getOutgoingDeps(current)
        : this.getIncomingDeps(current);

      for (const dep of deps) {
        const isExternal = this.fileResolver.isExternal(dep.target);

        if (isExternal && !includeExternal) continue;

        const resolvedTarget = isExternal
          ? dep.target
          : this.fileResolver.resolveImportTarget(current, dep.target);

        if (!resolvedTarget) continue;

        // Cycle detection
        if (currentPath.includes(resolvedTarget)) {
          cycles.push([...currentPath, resolvedTarget]);
          continue;
        }

        if (!visited.has(resolvedTarget)) {
          visited.add(resolvedTarget);
          results.push({
            file: resolvedTarget,
            depth: currentDepth + 1,
            importedSymbols: dep.symbols,
            isExternal,
          });

          if (!isExternal && currentDepth + 1 < maxDepth) {
            queue.push({
              file: resolvedTarget,
              depth: currentDepth + 1,
              path: [...currentPath, resolvedTarget],
            });
          }
        }
      }
    }

    return { results, cycles };
  }

  private getOutgoingDeps(filePath: string): Array<{ target: string; symbols: string[] }> {
    const rows = this.db.prepare(`
      SELECT target_symbol, metadata
      FROM relationships
      WHERE file_path = ? AND kind = 'imports'
      ORDER BY line
    `).all(filePath) as { target_symbol: string; metadata: string | null }[];

    // Group by module
    const grouped = new Map<string, string[]>();
    for (const row of rows) {
      const module = this.extractModule(row.target_symbol);
      if (!grouped.has(module)) grouped.set(module, []);
      const symbol = this.extractSymbolName(row.target_symbol);
      if (symbol) grouped.get(module)!.push(symbol);
    }

    return Array.from(grouped.entries()).map(([target, symbols]) => ({ target, symbols }));
  }

  private getIncomingDeps(filePath: string): Array<{ target: string; symbols: string[] }> {
    const basename = path.basename(filePath, path.extname(filePath));
    const rows = this.db.prepare(`
      SELECT DISTINCT file_path, target_symbol
      FROM relationships
      WHERE kind = 'imports'
        AND (target_symbol LIKE ? OR target_symbol LIKE ? OR target_symbol LIKE ?)
    `).all(`%/${basename}`, `%${basename}%`, filePath) as { file_path: string; target_symbol: string }[];

    const grouped = new Map<string, string[]>();
    for (const row of rows) {
      if (row.file_path === filePath) continue; // Skip self
      if (!grouped.has(row.file_path)) grouped.set(row.file_path, []);
      grouped.get(row.file_path)!.push(this.extractSymbolName(row.target_symbol) || '*');
    }

    return Array.from(grouped.entries()).map(([target, symbols]) => ({ target, symbols }));
  }

  private extractModule(targetSymbol: string): string {
    // target_symbol format: "module/path.symbol" or "module/path"
    const lastDot = targetSymbol.lastIndexOf('.');
    if (lastDot > 0 && !targetSymbol.includes('/')) return targetSymbol;
    if (lastDot > 0) return targetSymbol.substring(0, lastDot);
    return targetSymbol;
  }

  private extractSymbolName(targetSymbol: string): string {
    const lastDot = targetSymbol.lastIndexOf('.');
    if (lastDot > 0 && lastDot < targetSymbol.length - 1) {
      return targetSymbol.substring(lastDot + 1);
    }
    return path.basename(targetSymbol);
  }

  private mergeResults(outgoing: DependencyNode[], incoming: DependencyNode[]): DependencyNode[] {
    const seen = new Set<string>();
    const merged: DependencyNode[] = [];
    for (const node of [...outgoing, ...incoming]) {
      if (!seen.has(node.file)) {
        seen.add(node.file);
        merged.push(node);
      }
    }
    return merged;
  }

  private fileNotFoundResponse(file: string): DependencyResult {
    return {
      root: file,
      direction: 'outgoing',
      results: [],
      cycles: [],
      metadata: { totalNodes: 0, maxDepthReached: 0, truncated: false, queryTimeMs: 0, externalCount: 0 },
    };
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/graph/dependency_graph_service.py
"""Dependency Graph Service — BFS traversal on import relationships. KSA-179.

Uses NetworkX DiGraph for efficient graph traversal and cycle detection.
"""

from __future__ import annotations

import posixpath
import sqlite3
import time
from collections import deque

import networkx as nx

from .file_resolver import FileResolver
from .models import DependencyMetadata, DependencyNode, DependencyResult


class DependencyGraphService:
    """BFS traversal on import relationships with cycle detection.

    Supports outgoing (what does this file import?) and incoming (who imports this file?).
    """

    def __init__(self, conn: sqlite3.Connection, file_resolver: FileResolver) -> None:
        self._conn = conn
        self._file_resolver = file_resolver

    def query(
        self,
        file: str,
        direction: str = "outgoing",
        depth: int = 1,
        include_external: bool = False,
        limit: int = 50,
    ) -> DependencyResult:
        """Query dependency graph with direction and depth control."""
        start_time = time.time()
        clamped_depth = min(max(depth, 1), 5)

        resolved = self._file_resolver.resolve_file(file)
        if not resolved:
            return self._file_not_found(file)

        if direction == "both":
            out = self._bfs_traversal(resolved, "outgoing", clamped_depth, include_external, limit)
            inc = self._bfs_traversal(resolved, "incoming", clamped_depth, include_external, limit)
            results = self._merge_results(out["results"], inc["results"])
            cycles = out["cycles"] + inc["cycles"]
        else:
            traversal = self._bfs_traversal(resolved, direction, clamped_depth, include_external, limit)
            results = traversal["results"]
            cycles = traversal["cycles"]

        elapsed_ms = int((time.time() - start_time) * 1000)
        max_d = max((r.depth for r in results), default=0)

        return DependencyResult(
            root=resolved,
            direction=direction,
            results=results,
            cycles=cycles,
            metadata=DependencyMetadata(
                total_nodes=len(results),
                max_depth_reached=min(clamped_depth, max_d),
                truncated=len(results) >= limit,
                query_time_ms=elapsed_ms,
                external_count=sum(1 for r in results if r.is_external),
            ),
        )

    def build_networkx_graph(self, root: str, direction: str, depth: int) -> nx.DiGraph:
        """Build a NetworkX DiGraph from the dependency data for advanced analysis."""
        G = nx.DiGraph()
        G.add_node(root, depth=0)

        result = self.query(root, direction, depth, include_external=False, limit=200)
        for node in result.results:
            G.add_node(node.file, depth=node.depth, is_external=node.is_external)
            if direction == "outgoing":
                G.add_edge(root if node.depth == 1 else node.file, node.file,
                           symbols=node.imported_symbols)
            else:
                G.add_edge(node.file, root if node.depth == 1 else node.file,
                           symbols=node.imported_symbols)

        return G

    def detect_cycles_nx(self, root: str, depth: int = 3) -> list[list[str]]:
        """Use NetworkX to detect all simple cycles reachable from root."""
        G = self.build_networkx_graph(root, "outgoing", depth)
        try:
            return [list(c) for c in nx.simple_cycles(G)]
        except nx.NetworkXError:
            return []

    def _bfs_traversal(
        self,
        root: str,
        direction: str,
        max_depth: int,
        include_external: bool,
        limit: int,
    ) -> dict:
        visited: set[str] = {root}
        results: list[DependencyNode] = []
        cycles: list[list[str]] = []
        queue: deque[tuple[str, int, list[str]]] = deque()
        queue.append((root, 0, [root]))

        while queue and len(results) < limit:
            current, current_depth, current_path = queue.popleft()
            if current_depth >= max_depth:
                continue

            deps = (
                self._get_outgoing_deps(current)
                if direction == "outgoing"
                else self._get_incoming_deps(current)
            )

            for dep in deps:
                is_external = self._file_resolver.is_external(dep["target"])
                if is_external and not include_external:
                    continue

                resolved_target = (
                    dep["target"]
                    if is_external
                    else self._file_resolver.resolve_import_target(current, dep["target"])
                )
                if not resolved_target:
                    continue

                # Cycle detection
                if resolved_target in current_path:
                    cycles.append([*current_path, resolved_target])
                    continue

                if resolved_target not in visited:
                    visited.add(resolved_target)
                    results.append(DependencyNode(
                        file=resolved_target,
                        depth=current_depth + 1,
                        imported_symbols=dep["symbols"],
                        is_external=is_external,
                    ))

                    if not is_external and current_depth + 1 < max_depth:
                        queue.append((resolved_target, current_depth + 1, [*current_path, resolved_target]))

        return {"results": results, "cycles": cycles}

    def _get_outgoing_deps(self, file_path: str) -> list[dict]:
        cur = self._conn.execute(
            """SELECT target_symbol, metadata
               FROM relationships
               WHERE file_path = ? AND kind = 'imports'
               ORDER BY line""",
            (file_path,),
        )
        grouped: dict[str, list[str]] = {}
        for row in cur.fetchall():
            module = self._extract_module(row[0])
            if module not in grouped:
                grouped[module] = []
            symbol = self._extract_symbol_name(row[0])
            if symbol:
                grouped[module].append(symbol)

        return [{"target": t, "symbols": s} for t, s in grouped.items()]

    def _get_incoming_deps(self, file_path: str) -> list[dict]:
        basename = posixpath.basename(file_path).rsplit(".", 1)[0]
        cur = self._conn.execute(
            """SELECT DISTINCT file_path, target_symbol
               FROM relationships
               WHERE kind = 'imports'
                 AND (target_symbol LIKE ? OR target_symbol LIKE ? OR target_symbol LIKE ?)""",
            (f"%/{basename}", f"%{basename}%", file_path),
        )
        grouped: dict[str, list[str]] = {}
        for row in cur.fetchall():
            if row[0] == file_path:
                continue
            if row[0] not in grouped:
                grouped[row[0]] = []
            grouped[row[0]].append(self._extract_symbol_name(row[1]) or "*")

        return [{"target": t, "symbols": s} for t, s in grouped.items()]

    @staticmethod
    def _extract_module(target_symbol: str) -> str:
        last_dot = target_symbol.rfind(".")
        if last_dot > 0 and "/" not in target_symbol:
            return target_symbol
        if last_dot > 0:
            return target_symbol[:last_dot]
        return target_symbol

    @staticmethod
    def _extract_symbol_name(target_symbol: str) -> str:
        last_dot = target_symbol.rfind(".")
        if 0 < last_dot < len(target_symbol) - 1:
            return target_symbol[last_dot + 1:]
        return posixpath.basename(target_symbol)

    @staticmethod
    def _merge_results(outgoing: list[DependencyNode], incoming: list[DependencyNode]) -> list[DependencyNode]:
        seen: set[str] = set()
        merged: list[DependencyNode] = []
        for node in [*outgoing, *incoming]:
            if node.file not in seen:
                seen.add(node.file)
                merged.append(node)
        return merged

    @staticmethod
    def _file_not_found(file: str) -> DependencyResult:
        return DependencyResult(root=file, direction="outgoing")
